### PR TITLE
[p4fmt]: Add a pretty printer for p4fmt. 

### PR DESCRIPTION
**Pull request by snapdgn**
pretty printer for p4fmt

### CODE REVIEW

**Review by fruffy — APPROVED** [review]

**Review by fruffy — COMMENTED** [review]
I would call it a different name. P4Formatter? Also feel free to prune as many members and methods as possible. I think most of them are not needed here.

Also add a preorder function for `const IR::Node *` throw a P4C_UNIMPLEMENTED exception there.

**Comment by snapdgn** [comment]
> Also add a preorder function for `const IR::Node *` throw a P4C_UNIMPLEMENTED exception there.

Yes, was planning to do this.

File: backends/p4fmt/toP4/toP4.h [hunk 1]
```diff
@@ -0,0 +1,266 @@
+#ifndef P4FMT_TOP4_TOP4_H_
+#define P4FMT_TOP4_TOP4_H_
+
+#include "frontends/common/resolveReferences/resolveReferences.h"
+#include "ir/ir.h"
+#include "ir/visitor.h"
+#include "lib/sourceCodeBuilder.h"
+
+namespace P4Fmt {
+namespace P4 {
```
Comment (qobilidop) [inline backends/p4fmt/toP4/toP4.h:10]: I feel it's better not to introduce new nested namespaces unless absolutely necessary. Also there might be ambiguity between `::P4Fmt::P4` and `::P4` in unintentional ways. For now, put all P4 formatter code under `::P4Fmt` should be good enough.

**Comment by snapdgn** [comment]
> I would call it a different name. P4Formatter?

Just to clarify things, did you mean I should rename 'toP4' class to 'P4Formatter' ?

**Comment by fruffy** [comment]
> > I would call it a different name. P4Formatter?
> 
> Just to clarify things, did you mean I should rename 'toP4' class to 'P4Formatter' ?

Yes, or whatever name you find most appropriate.

**Review by fruffy — APPROVED** [review]

**Comment by fruffy** [comment]
Need to update the namespace after #4825 has been merged, then we are good to go.

**Comment by fruffy** [comment]
@snapdgn is this ready to be merged?

**Comment by snapdgn** [comment]
> @snapdgn is this ready to be merged? It is still in draft mode.

Sorry, this should be ready now.

File: backends/p4fmt/p4formatter/p4formatter.h
Comment (qobilidop) [inline backends/p4fmt/p4formatter/p4formatter.h:1]: Is this nested directory `p4formatter/p4formatter.h` necessary? How about simply `p4formatter.h`?

**Review by qobilidop — APPROVED** [review]
LGTM other than the directory structure.

File: backends/p4fmt/p4formatter/p4formatter.h
Comment (snapdgn) [reply backends/p4fmt/p4formatter/p4formatter.h:1]: I agree. I'll do away with the nested dir.

File: backends/p4fmt/p4formatter.cpp [hunk 2]
```diff
@@ -0,0 +1,1515 @@
...
+    builder.spc();
+    builder.append(t->name);
+    builder.endOfStatement();
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Type_BaseList *t) {
+    builder.append("tuple<");
+    bool first = true;
+    for (auto a : t->components) {
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:206]: This looks like a common pattern (visit some collection with comma separator and this `fist` /  not-`first`). Consider refactoring into some common code of visitation with separator. E.g. you can submit a lambda to visit the internals, e.g.:
```c++
visit_collection(t->components, ",", 
                 [](const auto *elem) {
                      auto p4type = a->getP4Type();
                      CHECK_NULL(p4type);
                      visit(p4type);
                   });
```

File: backends/p4fmt/p4formatter.cpp [hunk 3]
```diff
@@ -0,0 +1,1515 @@
...
+    }
+    builder.append("enum ");
+    visit(t->type);
+    builder.spc();
+    builder.append(t->name);
+    builder.spc();
+    builder.blockStart();
+    bool first = true;
+    for (auto a : t->members) {
+        if (!first) builder.append(",\n");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:270]: See above

File: backends/p4fmt/p4formatter.cpp [hunk 4]
```diff
@@ -0,0 +1,1515 @@
...
+    }
+    builder.newline();
+    builder.blockEnd(true);
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::TypeParameters *t) {
+    if (!t->empty()) {
+        builder.append("<");
+        bool first = true;
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:285]: ditto

File: backends/p4fmt/p4formatter.cpp [hunk 5]
```diff
@@ -0,0 +1,1515 @@
...
+        }
+        builder.appendFormat("%s ", name);
+    }
+    builder.append(t->name);
+    visit(t->typeParameters);
+    if (!isDeclaration) return false;
+    builder.spc();
+    builder.blockStart();
+
+    std::map<const IR::StructField *, cstring> type;
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:417]: `unordered_map`

File: backends/p4fmt/p4formatter.cpp [hunk 6]
```diff
@@ -0,0 +1,1515 @@
...
+    }
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Declaration_MatchKind *d) {
+    builder.append("match_kind ");
+    builder.blockStart();
+    bool first = true;
+    for (auto a : *d->getDeclarations()) {
+        if (!first) builder.append(",\n");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:598]: ditto

File: backends/p4fmt/p4formatter.cpp [hunk 7]
```diff
@@ -0,0 +1,1515 @@
...
+    doneVec();
+    builder.blockEnd(true);
+    expressionPrecedence = prec;
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::ListExpression *e) {
+    using ::P4::literals::operator""_cs;
+    cstring start, end;
+    if (listTerminators.empty()) {
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:752]: can't you use structured bindings to simplify this?

File: backends/p4fmt/p4formatter.cpp [hunk 8]
```diff
@@ -0,0 +1,1515 @@
...
+        builder.append("(list<");
+        visit(e->elementType->getP4Type());
+        builder.append(">)");
+    }
+    builder.append("{");
+    int prec = expressionPrecedence;
+    expressionPrecedence = DBPrint::Prec_Low;
+    bool first = true;
+    for (auto c : e->components) {
+        if (!first) builder.append(",");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:784]: see above

File: backends/p4fmt/p4formatter.cpp [hunk 9]
```diff
@@ -0,0 +1,1515 @@
...
+        builder.append("(");
+        visit(e->structType);
+        builder.append(")");
+    }
+    builder.append("{");
+    int prec = expressionPrecedence;
+    expressionPrecedence = DBPrint::Prec_Low;
+    bool first = true;
+    for (auto c : e->components) {
+        if (!first) builder.append(",");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:813]: ditto

File: backends/p4fmt/p4formatter.cpp [hunk 10]
```diff
@@ -0,0 +1,1515 @@
...
+        builder.append("(");
+        visit(e->headerStackType);
+        builder.append(")");
+    }
+    builder.append("{");
+    int prec = expressionPrecedence;
+    expressionPrecedence = DBPrint::Prec_Low;
+    bool first = true;
+    for (auto c : e->components) {
+        if (!first) builder.append(",");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:835]: ditto

File: backends/p4fmt/p4formatter.cpp [hunk 11]
```diff
@@ -0,0 +1,1515 @@
...
+        setVecSep(", ");
+        visit(e->typeArguments);
+        doneVec();
+        builder.append(">");
+        isDeclaration = decl;
+    }
+    builder.append("(");
+    setVecSep(", ");
+    expressionPrecedence = DBPrint::Prec_Low;
+    withinArgument = true;
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:903]: again great example of possible RAII usage

File: backends/p4fmt/p4formatter.cpp [hunk 12]
```diff
@@ -0,0 +1,1515 @@
...
+
+bool P4Formatter::preorder(const IR::ForStatement *s) {
+    if (!s->annotations->annotations.empty()) {
+        visit(s->annotations);
+        builder.spc();
+    }
+    builder.append("for (");
+    bool first = true;
+    for (auto *d : s->init) {
+        if (!first) builder.append(", ");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1087]: see above

File: backends/p4fmt/p4formatter.cpp [hunk 13]
```diff
@@ -0,0 +1,1515 @@
...
+    doneVec();
+    builder.blockEnd(false);
+    return false;
+}
+
+////////////////////////////////////
+
+bool P4Formatter::preorder(const IR::Annotations *a) {
+    bool first = true;
+    for (const auto *anno : a->annotations) {
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1186]: ditto

File: backends/p4fmt/p4formatter.cpp [hunk 14]
```diff
@@ -0,0 +1,1515 @@
...
+        setVecSep(", ");
+        preorder(&a->expr);
+        doneVec();
+        builder.append(close);
+    }
+    if (!a->kv.empty()) {
+        builder.append(open);
+        bool first = true;
+        for (auto kvp : a->kv) {
+            if (!first) builder.append(", ");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1213]: ditto

File: backends/p4fmt/p4formatter.cpp [hunk 15]
```diff
@@ -0,0 +1,1515 @@
...
+        builder.append("[]");
+    }
+    if (!a->body.empty() && a->expr.empty() && a->kv.empty()) {
+        // Have an unparsed annotation.
+        // We could be prettier here with smarter logic, but let's do the easy
+        // thing by separating every token with a space.
+        builder.append(open);
+        bool first = true;
+        for (auto tok : a->body) {
+            if (!first) builder.append(" ");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1231]: ditto

File: backends/p4fmt/p4formatter.cpp [hunk 16]
```diff
@@ -0,0 +1,1515 @@
...
+            first = false;
+
+            bool haveStringLiteral =
+                tok->token_type == ::P4::P4Parser::token_type::TOK_STRING_LITERAL;
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1235]: do not use top-level qualified P4 namespace

File: backends/p4fmt/p4formatter.cpp [hunk 17]
```diff
@@ -0,0 +1,1515 @@
...
+    builder.newline();
+    builder.blockEnd(true);
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::ParameterList *p) {
+    builder.append("(");
+    bool first = true;
+    for (auto param : *p->getEnumerator()) {
+        if (!first) builder.append(", ");
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1304]: see above

File: backends/p4fmt/p4formatter.cpp [hunk 18]
```diff
@@ -0,0 +1,1515 @@
...
+    preorder(&v->actionList);
+    doneVec();
+    builder.blockEnd(false);
+    return false;
+}
+
+bool P4Formatter::preorder(const IR::Key *v) {
+    builder.blockStart();
+
+    std::map<const IR::KeyElement *, cstring> kf;
```
Comment (asl) [inline backends/p4fmt/p4formatter.cpp:1402]: `unordered_map`

**Review by asl — DISMISSED** [review]
See comments

File: backends/p4fmt/p4formatter.h [hunk 19]
```diff
@@ -0,0 +1,301 @@
+#ifndef BACKENDS_P4FMT_P4FORMATTER_H_
+#define BACKENDS_P4FMT_P4FORMATTER_H_
+
+#include "frontends/common/resolveReferences/resolveReferences.h"
+#include "ir/ir.h"
+#include "ir/visitor.h"
+#include "lib/sourceCodeBuilder.h"
+
+namespace P4::P4Fmt {
+using std::filesystem::path;
```
Comment (fruffy) [inline backends/p4fmt/p4formatter.h:10]: You do not need this, just use the fully qualified path.

File: backends/p4fmt/p4formatter.h [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+    }
+
+    bool isSystemFile(std::filesystem::path &file);
+    std::filesystem::path ifSystemFile(
+        const IR::Node *node);  // return file containing node if system file
+
+ public:
+    // Output is constructed here
+    Util::SourceCodeBuilder &builder;
+    /* FIXME  -- simplify this by getting rid of the 'builder' object and just emitting
```
Comment (fruffy) [inline backends/p4fmt/p4formatter.h:107]: Nit: `//` instead of `/*`

File: backends/p4fmt/p4formatter.h [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+          isDeclaration(true),
+          withinArgument(false),
+          builder(*new Util::SourceCodeBuilder()),
+          outStream(&std::cout),
+          mainFile(nullptr) {
+        visitDagOnce = false;
+        setName("P4Formatter");
+    }
+
+    using Inspector::preorder;
```
Comment (fruffy) [inline backends/p4fmt/p4formatter.h:149]: This is inherited from ToP4 but should not be needed.

**Review by fruffy — APPROVED** [review]
@asl I will go ahead and merge this today if you do not have any other comments.

**Review by qobilidop — APPROVED** [review]

**Comment by fruffy** [comment]
Dismissing @asl's review to merge, we can address remaining issues in subsequent PRs.

**Comment by snapdgn** [comment]
> Most comments have been addressed.

Yes, I don't think I missed resolving any of the suggestions.

**Pull request merged**